src/Raven.Client/Documents/Session/AsyncDocumentSession.cs

```diff
@@ -91,13 +91,14 @@ public async Task RefreshAsync<T>(IEnumerable<T> entities, CancellationToken tok
                 var command = new GetDocumentsCommand(ids.ToArray(), includes: null, metadataOnly: false);
                 await RequestExecutor.ExecuteAsync(command, Context, sessionInfo: _sessionInfo, token).ConfigureAwait(false);
 
-                var j = 0;
-                foreach (var entity in entities)
+                var numOfCmdResults = command.Result.Results.Length;
+
+                for (int i = 0; i < numOfCmdResults; i++)
                 {
-                    var commandResult = (BlittableJsonReaderObject)command.Result.Results[j];
-                    RefreshInternal(entity, commandResult, documentInfos[j]);
-                    j++;
+                    var commandResult = (BlittableJsonReaderObject)command.Result.Results[i];
+                    RefreshInternal(entities.ElementAt(i), commandResult, documentInfos[i]);
```

Comment:
doesn't look correct?

consider in `entities` following order

`e1, e1, e1, e1, e1, e2` - you will send `e1-id, e2-id`  to the server only. Consider constructing dictionary here

```
Dictionary<string, (object Entity, DocumentInfo Info)>
```

then you can loop through all the results, extract the `Id` using `GetMetadata` and `GetId` extension methods and update the values from dict.

Thoughts?

Reply:
Done